File: CreationalPatterns/Builder/Builder.py
from abc import ABC, abstractmethod
from typing import List

# ---------- Product ----------
class Sandwich:
    def __init__(self) -> None:
        self.parts: List[str] = []

    def add_part(self, part: str) -> None:
        self.parts.append(part)

    def list_parts(self) -> None:
        for part in self.parts:
            print(part)

# ---------- Builder ----------
class Builder(ABC):
    @abstractmethod
    def make_bread(self) -> None:
        pass

    @abstractmethod
    def make_sauce(self) -> None:
        pass

    @abstractmethod
    def make_ingredients(self) -> None:
        pass

# ---------- Concrete Builders ----------
class VegSandwichBuilder(Builder):
    def __init__(self) -> None:
        self.reset()

    def reset(self) -> None:
        self.sandwich = Sandwich()

    def get_product(self) -> Sandwich:
        product = self.sandwich
        self.reset()
        return product

    def make_bread(self) -> None:
        self.sandwich.add_part("Non-Gluten Bread")

    def make_sauce(self) -> None:
        self.sandwich.add_part("Low Calories Sauce")

    def make_ingredients(self) -> None:
        self.sandwich.add_part("Mozzarella, Tomatoes, Basil")

class ChickenBurgerBuilder(Builder):
    def __init__(self) -> None:
        self.reset()

    def reset(self) -> None:
        self.sandwich = Sandwich()

    def get_product(self) -> Sandwich:
        product = self.sandwich
        self.reset()
        return product

    def make_bread(self) -> None:
        self.sandwich.add_part("Sesame Bun")

    def make_sauce(self) -> None:
        self.sandwich.add_part("Spicy Mayo")

    def make_ingredients(self) -> None:
        self.sandwich.add_part("Grilled Chicken, Lettuce, Tomato")

class BeefBurgerBuilder(Builder):
    def __init__(self) -> None:
        self.reset()

    def reset(self) -> None:
        self.sandwich = Sandwich()

    def get_product(self) -> Sandwich:
        product = self.sandwich
        self.reset()
        return product

    def make_bread(self) -> None:
        self.sandwich.add_part("Classic Bun")

    def make_sauce(self) -> None:
        self.sandwich.add_part("BBQ Sauce")

    def make_ingredients(self) -> None:
        self.sandwich.add_part("Beef Patty, Cheddar, Pickles, Onion")

# ---------- Director ----------
class Director:
    def __init__(self) -> None:
        self.builder: Builder | None = None

    def set_builder(self, builder: Builder) -> None:
        self.builder = builder

    def build_product(self) -> None:
        if self.builder is None:
            raise RuntimeError("Builder is not set")
        self.builder.make_bread()
        self.builder.make_sauce()
        self.builder.make_ingredients()

# ---------- Client ----------
def client_side() -> None:
    director = Director()

    # Veg Sandwich
    veg_builder = VegSandwichBuilder()
    director.set_builder(veg_builder)
    director.build_product()
    veg_sandwich = veg_builder.get_product()
    print("Veg Sandwich:")
    veg_sandwich.list_parts()
    print()

    # Chicken Burger
    chicken_builder = ChickenBurgerBuilder()
    director.set_builder(chicken_builder)
    director.build_product()
    chicken_burger = chicken_builder.get_product()
    print("Chicken Burger:")
    chicken_burger.list_parts()
    print()

    # Beef Burger
    beef_builder = BeefBurgerBuilder()
    director.set_builder(beef_builder)
    director.build_product()
    beef_burger = beef_builder.get_product()
    print("Beef Burger:")
    beef_burger.list_parts()
    print()

if __name__ == "__main__":
    client_side()

File: CreationalPatterns/Builder/Builder.cpp
/**
 * @file SandwichBuilder.cpp
 * @brief Example of the Builder design pattern in C++.
 *
 * This example demonstrates the Builder pattern, which separates the construction
 * of a complex object (Sandwich) from its representation. Different builders
 * (Veg, Chicken, Beef) construct different types of sandwiches step by step.
 * The Director controls the building process using a Builder interface, allowing
 * clients to obtain the finished Sandwich without knowing the construction details.
 */

#include <iostream>
#include <string>
#include <memory>
#include <vector>

/**
 * @class Sandwich
 * @brief The Product class representing a sandwich.
 */
class Sandwich
{
public:
    /// List all the parts of the sandwich
    void listParts()
    {
        for (const auto& part : parts) {
            std::cout << part << std::endl;
        }
    }

    /// Add a part to the sandwich
    void addPart(std::string part)
    {
        parts.push_back(std::move(part));
    }

private:
    std::vector<std::string> parts; ///< List of sandwich parts
};

/**
 * @class Builder
 * @brief Abstract Builder interface defining steps to build a sandwich.
 */
class Builder
{
public:
    virtual ~Builder() = default;
    virtual void makeBread() = 0;
    virtual void makeSauce() = 0;
    virtual void makeIngredients() = 0;
};

/**
 * @class VegSandwichBuilder
 * @brief Concrete Builder for a vegetarian sandwich.
 */
class VegSandwichBuilder : public Builder
{
public:
    VegSandwichBuilder() { reset(); }
    ~VegSandwichBuilder() = default;

    /// Reset the internal sandwich object
    void reset() {
        sandwich = std::make_unique<Sandwich>();
    }

    /// Return the finished sandwich
    std::unique_ptr<Sandwich> getProduct() {
        auto res{std::move(sandwich)};
        reset();
        return res;
    }

    void makeBread() override
    {
        sandwich->addPart("Whole Grain Bread");
    }

    void makeSauce() override 
    {
        sandwich->addPart("Low Calories Hummus");
    }

    void makeIngredients() override
    {
        sandwich->addPart("Mozzarella, Tomatoes, Cucumber, Lettuce");
    }
private:
    std::unique_ptr<Sandwich> sandwich;
};

/**
 * @class Director
 * @brief Director class that constructs a sandwich using a builder.
 */
class Director
{
public:
    void setBuilder(Builder* b) {
        builder = b;
    }

    void buildProduct()
    {
        if (!builder) throw std::runtime_error("Builder is not set");

        builder->makeBread();
        builder->makeSauce();
        builder->makeIngredients();
    }
private:
    Builder* builder{nullptr};
};

/**
 * @class ChickenBurgerBuilder
 * @brief Concrete Builder for a chicken burger.
 */
class ChickenBurgerBuilder : public Builder {
public:
    ChickenBurgerBuilder() { reset(); }

    void reset() { sandwich = std::make_unique<Sandwich>(); }

    std::unique_ptr<Sandwich> getProduct() {
        auto res = std::move(sandwich);
        reset();
        return res;
    }

    void makeBread() override {
        sandwich->addPart("Sesame Bun");
    }

    void makeSauce() override {
        sandwich->addPart("Spicy Mayo");
    }

    void makeIngredients() override {
        sandwich->addPart("Grilled Chicken, Lettuce, Tomato");
    }

private:
    std::unique_ptr<Sandwich> sandwich;
};

/**
 * @class BeefBurgerBuilder
 * @brief Concrete Builder for a beef burger.
 */
class BeefBurgerBuilder : public Builder {
public:
    BeefBurgerBuilder() { reset(); }

    void reset() { sandwich = std::make_unique<Sandwich>(); }

    std::unique_ptr<Sandwich> getProduct() {
        auto res = std::move(sandwich);
        reset();
        return res;
    }

    void makeBread() override {
        sandwich->addPart("Classic Bun");
    }

    void makeSauce() override {
        sandwich->addPart("BBQ Sauce");
    }

    void makeIngredients() override {
        sandwich->addPart("Beef Patty, Cheddar, Pickles, Onion");
    }

private:
    std::unique_ptr<Sandwich> sandwich;
};

/**
 * @brief Client code demonstrating the Builder pattern with different sandwiches
 */
void clientSide() {
    Director director;

    // Veg Sandwich
    VegSandwichBuilder vegBuilder;
    director.setBuilder(&vegBuilder);
    director.buildProduct();
    auto vegSandwich = vegBuilder.getProduct();
    std::cout << "Veg Sandwich:\n";
    vegSandwich->listParts();
    std::cout << std::endl;

    // Chicken Burger
    ChickenBurgerBuilder chickenBuilder;
    director.setBuilder(&chickenBuilder);
    director.buildProduct();
    auto chickenBurger = chickenBuilder.getProduct();
    std::cout << "Chicken Burger:\n";
    chickenBurger->listParts();
    std::cout << std::endl;

    // Beef Burger
    BeefBurgerBuilder beefBuilder;
    director.setBuilder(&beefBuilder);
    director.buildProduct();
    auto beefBurger = beefBuilder.getProduct();
    std::cout << "Beef Burger:\n";
    beefBurger->listParts();
    std::cout << std::endl;
}

int main()
{
    clientSide();
}
